src/coreclr/jit/morph.cpp

```diff
@@ -2544,6 +2502,51 @@ bool Compiler::fgTryMorphStructArg(CallArg* arg)
     return true;
 }
 
+//-----------------------------------------------------------------------------
+// FieldsMatchAbi:
+//   Check if the fields of a local map cleanly (in terms of offsets) to the
+//   specified ABI info.
+//
+// Arguments:
+//   varDsc  - promoted local
+//   abiInfo - ABI information
+//
+// Returns:
+//   True if it does. In that case FIELD_LIST usage is allowed for split args
+//   by the backend.
+//
+bool Compiler::FieldsMatchAbi(LclVarDsc* varDsc, const ABIPassingInformation& abiInfo)
+{
+    if (varDsc->lvFieldCnt != abiInfo.CountRegsAndStackSlots())
+    {
+        return false;
+    }
+
+    for (const ABIPassingSegment& seg : abiInfo.Segments())
+    {
+        if (seg.IsPassedInRegister())
+        {
+            unsigned fieldLclNum = lvaGetFieldLocal(varDsc, seg.Offset);
+            if (fieldLclNum == BAD_VAR_NUM)
+            {
+                return false;
+            }
+        }
+        else
+        {
+            for (unsigned offset = 0; offset < seg.Size; offset += TARGET_POINTER_SIZE)
+            {
+                if (lvaGetFieldLocal(varDsc, seg.Offset + offset) == BAD_VAR_NUM)
+                {
+                    return false;
+                }
+            }
+        }
+    }
+
+    return true;
+}
```

Comment:
This is only used for args split across registers and stack (arm32 mainly). These args retain the old IR requirements of needing fields to map cleanly to registers.

I might clean that up as a separate future PR, but I have a long term ambition to remove special handling of split arguments in the backend entirely by physically splitting the arg in the call node into multiple args instead, as part of lowering. In fact I think multireg args could also be handled this way which would make the backend much more uniform around this.